File: sdk/java-sdk/src/main/java/com/evernym/verity/sdk/protocols/connecting/v0_6/ConnectingV0_6.java
package com.evernym.verity.sdk.protocols.connecting.v0_6;

import com.evernym.verity.sdk.exceptions.UndefinedContextException;
import com.evernym.verity.sdk.exceptions.VerityException;
import com.evernym.verity.sdk.exceptions.WalletException;
import com.evernym.verity.sdk.protocols.MessageFamily;
import com.evernym.verity.sdk.protocols.connecting.v1_0.ConnectionsImplV1_0;
import com.evernym.verity.sdk.utils.Context;
import com.evernym.verity.sdk.utils.Util;
import org.json.JSONObject;

import java.io.IOException;

public interface ConnectingV0_6 extends MessageFamily {
    String QUALIFIER = Util.EVERNYM_MSG_QUALIFIER;
    String FAMILY = "connecting";
    String VERSION = "0.6";

    default String qualifier() {return QUALIFIER;}
    default String family() { return FAMILY;}
    default String version() {return VERSION;}

    String CREATE_CONNECTION = "CREATE_CONNECTION";
    String GET_STATUS = "get-status";

    static ConnectingImplV0_6 v0_6(String sourceId) {
        return new ConnectingImplV0_6(sourceId);
    }

    static ConnectingImplV0_6 v0_6(String sourceId, boolean includePublicDID) {
        return new ConnectingImplV0_6(sourceId, includePublicDID);
    }

    static ConnectingImplV0_6 v0_6(String sourceId, String phoneNo) {
        return new ConnectingImplV0_6(sourceId, phoneNo);
    }

    static ConnectingImplV0_6 v0_6(String sourceId, String phoneNo, boolean includePublicDID) {
        return new ConnectingImplV0_6(sourceId, phoneNo, includePublicDID);
    }

    static ConnectionsImplV1_0 v1_0(String parentThreadId, String label, String base64InviteURL) {
        return new ConnectionsImplV1_0(parentThreadId, label, base64InviteURL);
    }

    /**
     * Sends the connection create message to Verity
     *
     * @param context an instance of the Context object configured to a Verity Application
     * @throws IOException               when the HTTP library fails to post to the agency endpoint
     * @throws UndefinedContextException when the context doesn't have enough information for this operation
     * @throws WalletException when there are issues with encryption and decryption
     */
    void connect(Context context) throws IOException, VerityException;

    /**
     *
     * @return
     * @throws VerityException
     */
    JSONObject connectMsg(Context context) throws VerityException;

    /**
     *
     * @param context an instance of the Context object configured to a Verity Application
     * @return
     * @throws VerityException
     */
    byte[] connectMsgPacked(Context context) throws VerityException;

    /**
     * Sends the get status message to the connection
     * @param context an instance of the Context object configured to a Verity Application
     * @throws IOException               when the HTTP library fails to post to the agency endpoint
     * @throws UndefinedContextException when the context doesn't have enough information for this operation
     * @throws WalletException when there are issues with encryption and decryption
     */
    void status(Context context) throws IOException, VerityException;

    /**
     *
     * @return
     * @throws VerityException
     */
    JSONObject statusMsg(Context context) throws VerityException;

    /**
     *
     * @param context an instance of the Context object configured to a Verity Application
     * @return
     * @throws VerityException
     */
    byte[] statusMsgPacked(Context context) throws VerityException;

    void accept(Context context) throws IOException, VerityException;
    JSONObject acceptMsg(Context context) throws IOException, VerityException;
    byte[] acceptMsgPacked(Context context) throws IOException, VerityException;

}

File: sdk/nodejs-sdk/src/protocols/connection/ConnectingV06.js
'use strict'
const uuid = require('uuid')
const utils = require('../../utils')
const Protocol = require('../Protocol')

module.exports = class ConnectingV06 extends Protocol {
  constructor (threadId = null, sourceId = uuid(), phoneNumber = null, includePublicDID = false) {
    const msgFamily = 'connecting'
    const msgFamilyVersion = '0.6'
    const msgQualifier = utils.constants.EVERNYM_MSG_QUALIFIER
    super(msgFamily, msgFamilyVersion, msgQualifier, threadId)
    this.sourceId = sourceId
    this.phoneNumber = phoneNumber
    this.includePublicDID = includePublicDID

    this.msgNames.CREATE_CONNECTION = 'CREATE_CONNECTION'
    this.msgNames.INVITE_DETAIL = 'CONN_REQUEST_RESP'
    this.msgNames.CONN_REQ_ACCEPTED = 'CONN_REQ_ACCEPTED'
    this.msgNames.GET_STATUS = 'get-status'
  }

  async connectMsg (context) {
    const msg = this._getBaseMessage(this.msgNames.CREATE_CONNECTION)
    msg.sourceId = this.sourceId
    msg.phoneNo = this.phoneNo
    msg.includePublicDID = this.includePublicDID
    return msg
  }

  async connectMsgPacked (context) {
    return this.getMessageBytes(context, await this.connectMsg(context))
  }

  async connect (context) {
    await this.sendMessage(context, await this.connectMsgPacked(context))
  }

  async statusMsg (context) {
    const msg = this._getBaseMessage(this.msgNames.GET_STATUS)
    msg.sourceId = this.sourceId
    return msg
  }

  async statusMsgPacked (context) {
    return this.getMessageBytes(context, await this.statusMsg(context))
  }

  async status (context) {
    await this.sendMessage(context, await this.statusMsgPacked(context))
  }
}
